File: api-python/routers/amenities.py
from typing import List, Optional
from fastapi import APIRouter, HTTPException, Query, status
from database import get_db
from models import AmenitySuggestion
from utils.amenities import CANONICAL_AMENITIES, get_label_for_slug

router = APIRouter(prefix="/amenities", tags=["Amenities"])

@router.get("/suggest", response_model=List[AmenitySuggestion])
async def suggest_amenities(q: Optional[str] = None):
    if not q:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="The 'q' query parameter is required for autocomplete search."
        )

    db = get_db()

    # Aggregate occurrence counts of amenities across checkins
    pipeline = [
        {"$unwind": "$amenities_observed"},
        {"$group": {"_id": "$amenities_observed", "count": {"$sum": 1}}}
    ]

    cursor = db.checkins.aggregate(pipeline)
    aggregated_counts = await cursor.to_list(length=None)

    counts_map = {item["_id"]: item["count"] for item in aggregated_counts}

    # Construct complete suggestions list
    suggestions = []
    for item in CANONICAL_AMENITIES:
        slug = item["slug"]
        label = item["label"]
        suggestions.append({
            "slug": slug,
            "label": label,
            "usage_count": counts_map.get(slug, 0)
        })

    # Add dynamically observed user-defined amenities if they aren't in the canonical list
    for slug, count in counts_map.items():
        if not any(s["slug"] == slug for s in suggestions):
            suggestions.append({
                "slug": slug,
                "label": get_label_for_slug(slug),
                "usage_count": count
            })

    # Filter suggestions where query string matches slug or label case-insensitively
    query_lower = q.lower()
    filtered = [
        AmenitySuggestion(**s)
        for s in suggestions
        if query_lower in s["slug"].lower() or query_lower in s["label"].lower()
    ]

    # Sort by usage count descending, then alphabetically by label
    filtered.sort(key=lambda s: (-s.usage_count, s.label))

    return filtered

File: api-node/src/routes/amenities.ts
import { Router, Request, Response } from "express";
import { getCheckInsCollection } from "../db";
import { CANONICAL_AMENITIES, getLabelForSlug } from "../utils/amenities";

const router = Router();

// GET /amenities/suggest - Autocomplete endpoint
router.get("/suggest", async (req: Request, res: Response) => {
  try {
    const q = req.query.q as string | undefined;

    if (!q || typeof q !== "string") {
      return res.status(400).json({
        code: "BAD_REQUEST",
        message: "The 'q' query parameter is required for autocomplete search."
      });
    }

    // Run aggregation to fetch usage counts from all user check-ins
    const aggregatedCounts = await getCheckInsCollection()
      .aggregate([
        { $unwind: "$amenities_observed" },
        { $group: { _id: "$amenities_observed", count: { $sum: 1 } } }
      ])
      .toArray();

    // Map: amenity_slug -> count
    const countsMap: Record<string, number> = {};
    aggregatedCounts.forEach((item) => {
      countsMap[item._id] = item.count;
    });

    // Construct full suggestion pool from canonical list
    const allSuggestions = CANONICAL_AMENITIES.map((item) => ({
      slug: item.slug,
      label: item.label,
      usage_count: countsMap[item.slug] || 0
    }));

    // Dynamically include any additional user-defined amenities present in DB
    Object.keys(countsMap).forEach((slug) => {
      if (!allSuggestions.some((s) => s.slug === slug)) {
        allSuggestions.push({
          slug,
          label: getLabelForSlug(slug),
          usage_count: countsMap[slug]
        });
      }
    });

    // Filter by query matching slug or label
    const queryLower = q.toLowerCase();
    const suggestions = allSuggestions.filter(
      (item) =>
        item.slug.toLowerCase().includes(queryLower) ||
        item.label.toLowerCase().includes(queryLower)
    );

    // Sort by usage count descending, then alphabetically
    suggestions.sort((a, b) => b.usage_count - a.usage_count || a.label.localeCompare(b.label));

    return res.json(suggestions);
  } catch (error: any) {
    console.error("Error suggesting amenities:", error);
    return res.status(500).json({
      code: "INTERNAL_SERVER_ERROR",
      message: error.message || "An unexpected error occurred."
    });
  }
});

export default router;
